File: typescript/src/apis/iprocess_model_consumer_api.ts
import {Subscription} from '@essential-projects/event_aggregator_contracts';
import {IIdentity} from '@essential-projects/iam_contracts';

import {DataModels} from '../data_models/index';
import {Messages} from '../messages/index';

/**
 * The IProcessModelConsumerApi is used to retreive ProcessModels and start ProcessInstances.
 */
export interface IProcessModelConsumerApi {
  /**
   * Retrieves a list of all ProcessModels that the requesting user is
   * authorized to see.
   *
   * @async
   * @param identity The requesting users identity.
   * @returns        A Promise, which resolves with the ProcessModel list,
   *                 or rejects an error, in case the request failed.
   */
  getProcessModels(identity: IIdentity): Promise<DataModels.ProcessModels.ProcessModelList>;

  /**
   * Retrieves a ProcessModel by its ID.
   *
   * @async
   * @param identity       The requesting users identity.
   * @param processModelId The ID of the ProcessModel to retrieve.
   * @returns              A Promise, which resolves with the ProcessModel,
   *                       or rejects an error, in case the request failed.
   *                       This can happen, if the ProcessModel was not found,
   *                       or the user is not authorized to see it.
   */
  getProcessModelById(identity: IIdentity, processModelId: string): Promise<DataModels.ProcessModels.ProcessModel>;

  /**
   * Retrieves a ProcessModel by a ProcessInstanceID.
   *
   * @async
   * @param identity          The requesting users identity.
   * @param processInstanceId The ProcessInstanceID of the ProcessModel to retrieve.
   * @returns                 A Promise, which resolves with the ProcessModel,
   *                          or rejects an error, in case the request failed.
   *                          This can happen, if the ProcessModel was not found,
   *                          or the user is not authorized to see it.
   */
  getProcessModelByProcessInstanceId(identity: IIdentity, processInstanceId: string): Promise<DataModels.ProcessModels.ProcessModel>;

  /**
   * Starts a new instance of a ProcessModel with a specific ID.
   * Depending on the type of callback used, this function will resolve either
   * immediately after the ProcessInstance was started, or after it has reached
   * an EndEvent.
   * This can either be a specific EndEvent, or the first EndEvent encountered
   * during execution.
   *
   * @async
   * @param identity          The requesting users identity.
   * @param processModelId    The ID of the ProcessModel to retrieve.
   * @param payload           Contains parameters to pass to the ProcessInstance.
   *                          Can optionally define a CorrelationId to use.
   * @param startCallbackType The type of start callback use. Depending on the
   *                          value used, the function will either resolve right
   *                          after starting the ProcessInstance,
   *                          or after reaching an EndEvent.
   * @param startEventId      The ID of the StartEvent through which to start
   *                          the ProcessInstance.
   * @param endEventId        Contains the ID of the EndEvent that the
   *                          ProcessEngine should wait for, before resolving.
   *                          Works only in conjunction with the startCallbackType
   *                          "CallbackOnEndEventReached".
   * @returns                 A Promise, which resolves with the execution result,
   *                          or rejects an error, in case the request failed.
   *                          This can happen, if the ProcessModel was not found,
   *                          or the user is not authorized to see it, or if
   *                          the ProcessInstance was interrupted prematurely because of an error.
   */
  startProcessInstance(identity: IIdentity,
                       processModelId: string,
                       payload: DataModels.ProcessModels.ProcessStartRequestPayload,
                       startCallbackType: DataModels.ProcessModels.StartCallbackType,
                       startEventId?: string,
                       endEventId?: string,
                       processEndedCallback?: Messages.CallbackTypes.OnProcessEndedCallback,
                      ): Promise<DataModels.ProcessModels.ProcessStartResponsePayload>;

  /**
   * Retrieves the result of a specific ProcessModel within a Correlation.
   * This only works for ProcessInstances that have finished execution.
   *
   * @async
   * @param identity       The requesting users identity.
   * @param correlationId  The ID of the correlation for which to retrieve
   *                       the result.
   * @param processModelId The ID of the ProcessModel for which to retrieve
   *                       the result.
   * @returns              A Promise, which resolves with the correlation result,
   *                       or rejects an error, in case the request failed.
   *                       This can happen, if the ProcessModel or correlation
   *                       were not found, or the user is not authorized to see
   *                       the result.
   */
  getProcessResultForCorrelation(identity: IIdentity, correlationId: string, processModelId: string): Promise<Array<DataModels.CorrelationResult>>;

  /**
   * Gets all active ProcessInstances belonging to the given identity.
   *
   * @async
   * @param   identity The identity for which to get the ProcessInstances.
   * @returns          The list of ProcessInstances.
   */
  getProcessInstancesByIdentity(identity: IIdentity): Promise<Array<DataModels.ProcessInstance>>;

  /**
   * Executes a callback when a process started.
   *
   * @async
   * @param   identity      The requesting users identity.
   * @param   callback      The callback that will be executed when a new
   *                        ProcessInstance was started.
   *                        The message passed to the callback contains further
   *                        information about the started process.
   * @param   subscribeOnce Optional: If set to true, the Subscription will be
   *                        automatically disposed, after the notification was
   *                        received once.
   * @returns               The Subscription created by the EventAggregator.
   */
  onProcessStarted(
    identity: IIdentity,
    callback: Messages.CallbackTypes.OnProcessStartedCallback,
    subscribeOnce?: boolean,
  ): Promise<Subscription>;

  /**
   * Executes a callback when a new ProcessInstance for a given ProcessModelId
   * was started.
   *
   * @async
   * @param   identity       The requesting users identity.
   * @param   callback       The callback that will be executed when a new
   *                         ProcessInstance was started.
   *                         The message passed to the callback contains further
   *                         information about the started ProcessInstance.
   * @param   processModelId The ID of the ProcessModel for which to receive
   *                         notifications.
   * @param   subscribeOnce  Optional: If set to true, the Subscription will be
   *                         automatically disposed, after the notification was
   *                         received once.
   * @returns                The Subscription created by the EventAggregator.
   */
  onProcessWithProcessModelIdStarted(
    identity: IIdentity,
    callback: Messages.CallbackTypes.OnProcessStartedCallback,
    processModelId: string,
    subscribeOnce?: boolean,
  ): Promise<Subscription>;

  /**
   * Executes a callback when a ProcessInstance is terminated.
   *
   * @async
   * @param   identity      The requesting users identity.
   * @param   callback      The callback that will be executed when a
   *                        ProcessInstance is terminated.
   *                         The message passed to the callback contains further
   *                         information about the ProcessInstance terminated.
   * @param   subscribeOnce Optional: If set to true, the Subscription will be
   *                        automatically disposed, after the notification was
   *                        received once.
   * @returns               The Subscription created by the EventAggregator.
   */
  onProcessTerminated(
    identity: IIdentity,
    callback: Messages.CallbackTypes.OnProcessTerminatedCallback,
    subscribeOnce?: boolean,
  ): Promise<Subscription>;

  /**
   * Executes a callback when a ProcessInstance ends.
   *
   * @async
   * @param identity        The requesting users identity.
   * @param callback        The callback that will be executed when a
   *                        ProcessInstance is finished.
   *                        The message passed to the callback contains further
   *                        information about the finished ProcessInstance.
   * @param   subscribeOnce Optional: If set to true, the Subscription will be
   *                        automatically disposed, after the notification was
   *                        received once.
   * @returns               The Subscription created by the EventAggregator.
   */
  onProcessEnded(
    identity: IIdentity,
    callback: Messages.CallbackTypes.OnProcessEndedCallback,
    subscribeOnce?: boolean,
  ): Promise<Subscription>;
}

File: dotnet/src/apis/IProcessModelConsumerApi.cs
namespace ProcessEngine.ConsumerAPI.Contracts.APIs
{
    using System.Collections.Generic;
    using System.Threading.Tasks;

    using EssentialProjects.IAM.Contracts;
    using ProcessEngine.ConsumerAPI.Contracts.DataModel;
    using ProcessEngine.ConsumerAPI.Contracts.Exceptions;

    /// <summary>
    /// The IProcessModelConsumerApi is used to retreive ProcessModels and start ProcessInstances.
    /// </summary>
    public interface IProcessModelConsumerApi
    {
        /// <summary>
        /// Retrieves a list of all ProcessModels that the requesting user is authorized to see.
        /// </summary>
        /// <param name="identity">The requesting users identity.</param>
        /// <returns>A list of accessible ProcessModels.</returns>
        /// <exception cref="System.UnauthorizedAccessException">Thrown when the identity has insufficient rights to perform the operation.</exception>
        /// <exception cref="ProcessNotFoundException"> Thrown when the ProcessModel could not be found.</exception>
        Task<ProcessModelList> GetProcessModels(IIdentity identity);

        /// <summary>
        /// Retrieves a ProcessModel by its ID.
        /// </summary>
        /// <param name="identity">The requesting users identity.</param>
        /// <param name="processModelId">The ID of the ProcessModel to retrieve.</param>
        /// <returns>The retrieved ProcessModel.</returns>
        /// <exception cref="System.UnauthorizedAccessException">Thrown when the identity has insufficient rights to perform the operation.</exception>
        /// <exception cref="ProcessNotFoundException"> Thrown when the ProcessModel could not be found.</exception>
        Task<ProcessModel> GetProcessModelById(IIdentity identity, string processModelId);

        /// <summary>
        /// Retrieves a ProcessModel by a ProcessInstanceID.
        /// </summary>
        /// <param name="identity">The requesting users identity.</param>
        /// <param name="processInstanceId">The ID of the ProcessInstance for which to retrieve the ProcessModel.</param>
        /// <returns>The retrieved ProcessModel.</returns>
        /// <exception cref="System.UnauthorizedAccessException">Thrown when the identity has insufficient rights to perform the operation.</exception>
        /// <exception cref="ProcessNotFoundException"> Thrown when the ProcessModel could not be found.</exception>
        Task<ProcessModel> GetProcessModelByProcessInstanceId(IIdentity identity, string processInstanceId);

        /// <summary>
        /// Gets all active ProcessInstances belonging to the given identity.
        /// </summary>
        /// <param name="identity">The identity for which to get the ProcessInstances.</param>
        /// <returns>The list of ProcessInstances.</returns>
        /// <exception cref="System.UnauthorizedAccessException">Thrown when the identity has insufficient rights to perform the operation.</exception>
        /// <exception cref="ProcessNotFoundException"> Thrown when the ProcessModel could not be found.</exception>
        Task<IEnumerable<ProcessInstance>> GetProcessInstancesByIdentity(IIdentity identity);


        /// <summary>
        /// Starts an instance for a given ProcessDefinition. Process variables and correlation id may be supplied in the request payload.
        /// </summary>
        /// <returns>A set of informations about the started ProcessInstance; such as the used ProcessInstance ID or the assigned Correlation ID.</returns>
        /// <param name="identity">The requesting users <see cref="EssentialProjects.IAM.Contracts.IIdentity">identity</see>. Should usually contain an auth token.</param>
        /// <param name="processModelId">The ID of the ProcessDefinition to be retrieved.</param>
        /// <param name="startEventId">The ID of a specific StartEvent to start the process with.</param>
        /// <param name="processStartRequestPayload">The payload for the Start request. Contains variables for the ProcessInstance and optionally a CorrelationId.</param>
        /// <param name="callbackType"><see cref="StartCallbackType">Callback type</see></param>
        /// <param name="endEventId">The ID of the end event to wait for, when callbackType == StartCallbackType.CallbackOnEndEventReached.</param>
        Task<ProcessStartResponsePayload> StartProcessInstance<TInputValues>(
            IIdentity identity,
            string processModelId,
            string startEventId,
            ProcessStartRequestPayload<TInputValues> processStartRequestPayload,
            StartCallbackType callbackType = StartCallbackType.CallbackOnProcessInstanceCreated,
            string endEventId = "")
        where TInputValues : new();

        /// <summary>
        /// Gets the ProcessInstance results of a given correlation.
        /// </summary>
        /// <returns>The Correlation's results.</returns>
        /// <param name="identity">The requesting users <see cref="EssentialProjects.IAM.Contracts.IIdentity">identity</see>. Should usually contain an auth token.</param>
        /// <param name="correlationId">The ID of the Correlation for which to get the results.</param>
        /// <param name="processModelId">The ID of the ProcessDefinition for which to get the results.</param>
        /// <typeparam name="TPayload">The type that holds the definition for the Correlation result's payload.</typeparam>
        Task<IEnumerable<CorrelationResult<TPayload>>> GetProcessResultForCorrelation<TPayload>(
            IIdentity identity,
            string correlationId,
            string processModelId)
        where TPayload : new();
    }
}
